sycl/doc/extensions/proposed/sycl_ext_intel_fpga_datapath.asciidoc

```diff
@@ -0,0 +1,223 @@
+template <typename T, typename PropListT = sycl::ext::oneapi::properties<>>
+class fpga_datapath {
+  ...
+----
+
+`fpga_datapath` is a class template, parameterized by the type of the underlying
+allocation _T_ which encodes the size of the allocation.
+
+`fpga_datapath` initializes the underlying object `T` with the values passed into its constructor.
+
+[NOTE]
+====
+
+`fpga_datapath` has similar semantics to a reference wrapper. An fpga_datapath templated on a type T can implicitly convert to T, thereby allowing it to be used in most places that an object of type T could be used. The dot operator
+(`operator.`) cannot be overloaded, so a `get()` member is provided to allow a
+reference to be extracted directly when needed.
+====
+
+=== FPGA Datapath Interface
+
+The section below and the table following describe the constructors and member functions for `fpga_datapath`.
+
+[source,c++]
+----
+namespace sycl::ext::intel::experimental {
+
+  template <typename T>
+  class fpga_datapath {
+  public:
+    template <typename... S> constexpr fpga_datapath(S... args);
+    
+    // Access the underlying data
+    operator T &() noexcept;
+    operator const T &() const noexcept;
+
+    T &get() noexcept;
+    const T &get() const noexcept;
+
+    // Enable assignments from underlying type
+    fpga_datapath &operator=(const T &) noexcept;
+
+    // Note that there is no need for "fpga_datapath" to define member functions
+    // for operators like "++", "[]", "->", comparison, etc. Instead, the type
+    // "T" need only define these operators as non-member functions. Because
+    // there is an implicit conversion from "fpga_datapath" to "T&", the operations
+    // can be applied to objects of type "fpga_datapath<T>"
+  };
+
+} // namespace sycl::ext::intel::experimental
+----
+
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename ... S>
+constexpr fpga_datapath(S ... args);
+----
+|
+Constructs a `fpga_datapath` object, and implicit storage for `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T&() noexcept;
+operator const T&() const noexcept;
+----
+|
+Implicit conversion to a reference to the underlying `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+T& get() noexcept;
+const T& get() const noexcept;
+----
+|
+Returns a reference to the underlying `T`.
+
+|===
+
+=== Restrictions on creating fpga_datapath objects
+
+There are restrictions on how the application can create objects of type
+`fpga_datapath`. Applications that violate these restrictions are ill-formed.
+
+* The `fpga_datapath` variable can only have accesses from a single kernel. This happens when `device_global` is templated on a `fpga_datapath` type. 
+
+* The `fpga_datapath` variable cannot be templated on a `fpga_mem` object, as this is a conflict of requirements on how the variable should be implemented, on or off the datapath. Diagnostic is required.
```

Comment:
This simplistic check doesn't catch all the cases that don't make sense because one could easily have the fpga_mem wrapped in another object.  For example:

```
class foo {
  fpga_mem<int> a;
  int b;
};
...
q.single_task([=] {
  fpga_datapath<foo> c;
});
```

But addressing this comprehensively is complicated as you have to recursively inspect all the member of any type that an fpga_datapath is templated on.  I don't even think that could be done without compiler support which would preclude a purely C++-based implementation of the extension.  It would have to be implemented in either _every_ backend, even non-FPGA ones, which is inefficient, or in the frontend.

I don't think this effort is worth it.  Could we instead simply create a simple rule for how the two extensions interact?  For example, we could state that the more narrow annotation wins.  In my example above, that would mean that a would be implemented in memory while b would be implemented in the datapath.

Reply:
I have defined the behaviour of templating `fpga_mem` on `fpga_datapath` or vice versa. 

As for having arrays of `fpga_mem` i eventually concluded that it should be allowed. Since it seems arbitrary to prevent aggregates of a class. I don't think any special notes about it should be included in the SYCL spec. We can call out this behavior in the FPGA specific documentation.